openedx/core/djangolib/django_plugins.py

```diff
@@ -0,0 +1,360 @@
+from django.conf.urls import include, url
+from logging import getLogger
+from openedx.core.lib.plugins import PluginManager
+
+
+log = getLogger(__name__)
+
+
+# Name of the class attribute to put in the AppConfig class of the Plugin App.
+PLUGIN_APP_CLASS_ATTRIBUTE_NAME = u'plugin_app'
+
+
+# Name of the function that belongs in the plugin Django app's settings file.
+# The function should be defined as:
+#     def plugin_settings(settings):
+#         # enter code that should be injected into the given settings module.
+PLUGIN_APP_SETTINGS_FUNC_NAME = u'plugin_settings'
+
+
+class ProjectType(object):
+    """
+    The ProjectType enum defines the possible values for the Django Projects
+    that are available in the edx-platform. Plugin apps use these values to
+    declare explicitly which projects they are extending.
+    """
+    lms = u'lms.djangoapp'
+    cms = u'cms.djangoapp'
+
+
+class SettingsType(object):
+    """
+    The SettingsType enum defines the possible values for the settings files
+    that are available for extension in the edx-platform. Plugin apps use these
+    values (in addition to ProjectType) to declare explicitly which settings
+    (in the specified project) they are extending.
+
+    See https://github.com/edx/edx-platform/master/lms/envs/docs/README.rst for
+    further information on each Settings Type.
+    """
+    aws = u'aws'
+    common = u'common'
+    devstack = u'devstack'
+    test = u'test'
+
+
+class PluginSettings(object):
+    """
+    The PluginSettings enum defines dictionary field names (and defaults)
+    that can be specified by a Plugin App in order to configure the settings
+    that are injected into the project.
+    """
+    config = u'settings_config'
+    relative_path = u'relative_path'
+    DEFAULT_RELATIVE_PATH = u'settings'
+
+
+class PluginURLs(object):
+    """
+    The PluginURLs enum defines dictionary field names (and defaults) that can
+    be specified by a Plugin App in order to configure the URLs that are
+    injected into the project.
+    """
+    config = u'url_config'
+    app_name = u'app_name'
+    namespace = u'namespace'
+    regex = u'regex'
+    relative_path = u'relative_path'
+    DEFAULT_RELATIVE_PATH = u'urls'
+
+
+class DjangoAppRegistry(PluginManager):
+    """
+    The DjangoAppRegistry class encapsulates the functionality to enable
+    improved plugin support of Django apps.
+    """
+
+    @classmethod
+    def get_plugin_apps(cls, project_type):
+        """
+        Returns a list of all registered Plugin Apps, expected to be added to
+        the INSTALLED_APPS list for the given project_type.
+        """
+        plugin_apps = [
+            u'{module_name}.{class_name}'.format(
+                module_name=app_config.__module__,
+                class_name=app_config.__name__,
+            )
+            for app_config in cls._get_app_configs(project_type)
+            if getattr(app_config, PLUGIN_APP_CLASS_ATTRIBUTE_NAME, True)
```

Comment:
So discovered plugins are only skipped if they explicitly set `plugin_app = False` or such?  That seems a little odd.

Reply:
good catch.  I need to update this.  This is from a previous iteration of the code where explicit config settings weren't required.  I changed the default value to `False`.

Reply:
Ok, in that case remember to add `plugin_app` to the `openedx.core.djangoapps.schedules` app's configuration so it can be loaded.